lib/langchain/tool/weather.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+module Langchain::Tool
+  class Weather < Base
+    #
+    # A weather tool that gets current, forecast, or historical weather data
+    #
+    # Gem requirements:
+    #   gem "open-weather-ruby-client", "~> 0.3.0"
+    #   api_key: https://home.openweathermap.org/api_keys
+    #
+    # Usage:
+    # weather = Langchain::Tool::Weather.new(api_key: "YOUR_API_KEY")
+    # weather.execute(input: "Boston, current, imperial")
+    #
+
+    NAME = "weather"
+
+    description <<~DESC
+      Useful for getting current, forecast, or historical weather data
+  
+      The input to this tool should be a city name followed by the type of weather (current or forecast) you want to get, followed optionally by the units (imperial, metric, or standard)
+      Example usage:
+        Action Input: Boston, current, imperial
+        Weather: The current weather in Boston is temp 52.47, feels_like 51.78, temp_min 48.56, temp_max 55.22, pressure 1006, humidity 93
+    DESC
+
+    attr_reader :client, :units
+
+    def initialize(api_key:, units: "metric")
+      depends_on "open-weather-ruby-client"
+      require "open-weather-ruby-client"
+
+      OpenWeather::Client.configure do |config|
+        config.api_key = api_key
+        config.user_agent = "Langchainrb Ruby Client"
+      end
+
+      @client = OpenWeather::Client.new
+    end
+
+    # Returns weather for a city
+    # @param input [String] comma separated city, type of weather (current, forecast or historical), and unit (optional: imperial, metric, or standard)
+    # @return [String] Answer
+    def execute(input:)
+      Langchain.logger.info("[#{self.class.name}]".light_blue + ": Executing for \"#{input}\"")
+
+      city = input.split(", ").first
+      if input.count(",") > 1
+        type = input.split(", ")[0..-2][1]
+        units = input.split(", ").last
+      else
+        type = input.split(", ").last
+      end
```

Comment:
I think you can simplify this whole block with:
```ruby
input_array = input.split(",")
city, type, units = *input_array
```

Reply:
Tried but that breaks it even though looks like it should work.  Will look into it.

Reply:
Fixed.  Will wait for further input on wisdom of adding additional gem and google maps api key requirement.

Reply:
@mattlindsey What do we need to pass the lat/long for? Does the API not accept the actual place string?

Reply:
No.  I believe the `forecast` data cannot be provided without lat/long, and the gem cannot determine those itself.  Weird situation.

Reply:
@andreibondarev To be clear, the API (not gem) provides the following with the free plan (1000 calls per day), but forecasts are by lat/long, so the Tool could do the following for free if the consumer of the tool passes in lat/long to `execute`:

Current weather
Minute forecast for 1 hour
Hourly forecast for 48 hours
Daily forecast for 8 days
Government weather alerts
Historical weather data for any timestamp for 40+ years historical archive

Taken from https://openweathermap.org/api
History daily aggregated weather data for 40+ years historical archive